src/Features/Core/Portable/Formatting/FormattingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Formatting
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    internal class FormattingDiagnosticAnalyzer
+        : AbstractCodeStyleDiagnosticAnalyzer
+    {
+        public static readonly string ReplaceTextKey = nameof(ReplaceTextKey);
+
+        public static readonly ImmutableDictionary<string, string> RemoveTextProperties =
+            ImmutableDictionary.Create<string, string>().Add(ReplaceTextKey, "");
+
+        public FormattingDiagnosticAnalyzer()
+            : base(
+                IDEDiagnosticIds.FormattingDiagnosticId,
+                new LocalizableResourceString(nameof(FeaturesResources.Formatting_analyzer_title), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                new LocalizableResourceString(nameof(FeaturesResources.Formatting_analyzer_message), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SyntaxAnalysis;
+
+        public override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxTreeAction(AnalyzeSyntaxTree);
+
+        private void AnalyzeSyntaxTree(SyntaxTreeAnalysisContext context)
+        {
+            if (!(context.Options is WorkspaceAnalyzerOptions workspaceAnalyzerOptions))
+            {
+                return;
+            }
+
+            var options = context.Options.GetDocumentOptionSetAsync(context.Tree, context.CancellationToken).GetAwaiter().GetResult();
+            if (options == null)
+            {
+                return;
+            }
+
+            var workspace = workspaceAnalyzerOptions.Services.Workspace;
+            var formattingChanges = Formatter.GetFormattedTextChanges(context.Tree.GetRoot(context.CancellationToken), workspace, options, context.CancellationToken);
```

Comment:
you should set the no concurrency option here rather than putting that in the API itself. I guess you are setting that option since analyzer engine itself parallelize analysis and nested concurrency is actually get you worse result than not doing concurrency.

Reply:
> you should set the no concurrency option here rather than putting that in the API itself.

No; I made the synchronous calls behave like synchronous calls. Previously there were cases where they silently blocked the calling thread and performed the work on an unspecified number of additional threads. The new behavior matches the existing signatures and thus what the callers would already be expecting.